File: frontend/js/views/merton_default.js
// Merton structural default — distance to default + PD, via /calc/merton-default.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const num = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 4 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 3 }) + '%');
export async function renderMertonDefault(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.merton.h1.title">// MERTON DEFAULT MODEL</span></h1>
        <p class="muted small" data-i18n="view.merton.hint.intro">Treats equity as a call option on the firm's assets to derive the distance to default and the probability of default. DD = (ln(V/D) + (r − ½σ²)T) / (σ√T); PD = N(−DD). A structural approximation, not a credit rating.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.merton.h2.inputs">Firm inputs</h2>
        <form id="merton-form" class="inline-form">
            <label><span data-i18n="view.merton.label.v">Asset value ($)</span><input type="number" step="1" min="0" name="asset_value_usd" value="100" required></label>
            <label><span data-i18n="view.merton.label.d">Debt face ($)</span><input type="number" step="1" min="0" name="debt_face_usd" value="80" required></label>
            <label><span data-i18n="view.merton.label.sig">Asset volatility (decimal)</span><input type="number" step="0.01" min="0" name="asset_volatility" value="0.30" required></label>
            <label><span data-i18n="view.merton.label.r">Risk-free rate (decimal)</span><input type="number" step="0.005" name="risk_free_rate" value="0.05"></label>
            <label><span data-i18n="view.merton.label.t">Horizon (years)</span><input type="number" step="0.25" min="0" name="horizon_years" value="1"></label>
        </form></div><div id="merton-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#merton-form'); const n = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const gen = async () => {
        const body = { asset_value_usd: n('asset_value_usd'), debt_face_usd: n('debt_face_usd'), asset_volatility: n('asset_volatility'), risk_free_rate: n('risk_free_rate'), horizon_years: n('horizon_years') };
        try { const d = await api.calcMertonDefault(body); if (!viewIsCurrent(tok)) return; res(mount, d); }
        catch (e) { showToast(e.message || t('view.merton.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250); form.addEventListener('input', () => live()); form.addEventListener('submit', (e) => { e.preventDefault(); gen(); }); gen();
}
function res(mount, d) {
    const el = mount.querySelector('#merton-result');
    if (!d.valid) { el.innerHTML = `<p class="muted" data-i18n="view.merton.invalid">Invalid inputs (value, debt, vol, horizon must be positive).</p>`; applyUiI18n(el); return; }
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card neg"><div class="label" data-i18n="view.merton.card.pd">Probability of default</div><div class="value">${pct(d.probability_of_default_pct)}</div></div>
        <div class="card pos"><div class="label" data-i18n="view.merton.card.dd">Distance to default</div><div class="value">${num(d.distance_to_default)}σ</div></div>
        <div class="card"><div class="label" data-i18n="view.merton.card.lev">Leverage</div><div class="value">${pct(d.leverage_pct)}</div></div>
    </div></div>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/merton_default.rs
//! Merton structural credit model — treats a firm's equity as a call option on
//! its assets and derives the distance to default and the probability of default.
//! With asset value V, asset volatility σ, debt face D maturing in T years, and
//! rate r, the distance to default is `DD = (ln(V/D) + (r − ½σ²)T) / (σ√T)` and the
//! (risk-neutral) default probability is `PD = N(−DD)`. Reports DD, PD, and the
//! leverage. Pure compute; a structural approximation, not a rating. Not advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct MertonInput {
    /// Market value of the firm's assets.
    pub asset_value_usd: f64,
    /// Face value of debt due at the horizon.
    pub debt_face_usd: f64,
    /// Asset volatility (annual, decimal, e.g. 0.30).
    pub asset_volatility: f64,
    /// Risk-free rate (decimal).
    #[serde(default)]
    pub risk_free_rate: f64,
    /// Horizon in years.
    #[serde(default = "default_horizon")]
    pub horizon_years: f64,
}

fn default_horizon() -> f64 {
    1.0
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct MertonReport {
    /// Distance to default, in standard deviations.
    pub distance_to_default: f64,
    /// Probability of default over the horizon, percent.
    pub probability_of_default_pct: f64,
    /// Debt ÷ assets, percent.
    pub leverage_pct: f64,
    pub valid: bool,
}

fn norm_cdf(x: f64) -> f64 {
    0.5 * erfc(-x / std::f64::consts::SQRT_2)
}

/// Complementary error function (Abramowitz & Stegun 7.1.26, max err ~1.5e-7).
fn erfc(x: f64) -> f64 {
    let z = x.abs();
    let t = 1.0 / (1.0 + 0.5 * z);
    let tau = t
        * (-z * z - 1.26551223
            + t * (1.00002368
                + t * (0.37409196
                    + t * (0.09678418
                        + t * (-0.18628806
                            + t * (0.27886807
                                + t * (-1.13520398
                                    + t * (1.48851587
                                        + t * (-0.82215223 + t * 0.17087277)))))))))
        .exp();
    if x >= 0.0 {
        tau
    } else {
        2.0 - tau
    }
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

pub fn generate(i: &MertonInput) -> MertonReport {
    if i.asset_value_usd <= 0.0
        || i.debt_face_usd <= 0.0
        || i.asset_volatility <= 0.0
        || i.horizon_years <= 0.0
    {
        return MertonReport::default();
    }
    let sig = i.asset_volatility;
    let t = i.horizon_years;
    let dd = ((i.asset_value_usd / i.debt_face_usd).ln()
        + (i.risk_free_rate - 0.5 * sig * sig) * t)
        / (sig * t.sqrt());
    let pd = norm_cdf(-dd);
    MertonReport {
        distance_to_default: round4(dd),
        probability_of_default_pct: round4(pd * 100.0),
        leverage_pct: round4(i.debt_face_usd / i.asset_value_usd * 100.0),
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.02
    }

    fn base() -> MertonInput {
        MertonInput {
            asset_value_usd: 100.0,
            debt_face_usd: 80.0,
            asset_volatility: 0.30,
            risk_free_rate: 0.05,
            horizon_years: 1.0,
        }
    }

    #[test]
    fn distance_and_pd() {
        let d = generate(&base());
        assert!(d.valid);
        assert!(close(d.distance_to_default, 0.7605));
        assert!(close(d.probability_of_default_pct, 22.35));
        assert!(close(d.leverage_pct, 80.0));
    }

    #[test]
    fn lower_leverage_lower_pd() {
        let d = generate(&MertonInput { debt_face_usd: 50.0, ..base() });
        assert!(d.distance_to_default > 0.7605);
        assert!(d.probability_of_default_pct < 22.35);
    }

    #[test]
    fn higher_vol_higher_pd() {
        let d = generate(&MertonInput { asset_volatility: 0.60, ..base() });
        assert!(d.probability_of_default_pct > 22.35);
    }

    #[test]
    fn invalid_inputs() {
        assert!(!generate(&MertonInput { asset_volatility: 0.0, ..base() }).valid);
        assert!(!generate(&MertonInput { debt_face_usd: 0.0, ..base() }).valid);
    }
}
